### PR TITLE
body params passed without undefined values, formData tests

### PR DESCRIPTION
**Pull request by fazpu**
- Build is passing.

- In the next PR, I will try to refactor the generated code a little bit so that the classes extend a common class, which will contain the HttpParams and Body params creation methods.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "start": "ts-node src/index.ts",
     "prebuild": "rm -rf dist",
     "build": "tsc --outDir dist",
-    "postbuild": "npm run generate",
+    "prepublishOnly": "cd demo-app/client && npm i && npm run generate",
```
Comment (fazpu) [inline package.json]: this is the first time I am using the "prepublishOnly" tag. Please, when publishing the package, check it is working correctly.

File: package.json [hunk 2]
```diff
@@ -10,7 +10,7 @@
     "start": "ts-node src/index.ts",
     "prebuild": "rm -rf dist",
     "build": "tsc --outDir dist",
-    "postbuild": "npm run generate",
```
Comment (fazpu) [inline package.json]: when the "prepublishOnly" is implemented, the "postbuild" is unnecessary

File: src/requests/process-method.ts [hunk 3]
```diff
@@ -89,10 +89,26 @@ function getParamSeparation(paramGroups: Dictionary<Parameter[]>): string[] {
       return def;
     }
 
-    // only one direct body parameter is allowed
-    // 1st one taken if more of them present
     if (groupName === 'body') {
-      def = `params.${group[0].name};`;
+      // when the schema: { '$ref': '#/definitions/exampleDto' } construct is used
+      if ('schema' in group[0]) {
```
Comment (fazpu) [inline src/requests/process-method.ts]: if the parameters contain "schema" attribute, only the "schema" attribute will be applied

File: src/requests/process-method.ts [hunk 4]
```diff
@@ -64,6 +64,23 @@ export function processMethod(method: ControllerMethod): MethodOutput {
...
+function getParamsSignature(processedParams: ProcessParamsOutput, paramsType: string) {
+    return (!processedParams.isInterfaceEmpty) ? `params: ${paramsType}` : '';
+}
+
+/**
+ * Creates a definition of interfaceDef, which defines interface for the http method input
+ * @param processedParams
+ */
+function getInterfaceDef(processedParams: ProcessParamsOutput) {
+    return (!processedParams.isInterfaceEmpty) ? processedParams.paramDef : '';
```
Comment (jnwltr) [inline src/requests/process-method.ts]: Parentheses are not needed around 1st operand here and a couple places in the commit.

File: src/requests/process-method.ts [hunk 5]
```diff
@@ -89,10 +89,26 @@ function getParamSeparation(paramGroups: Dictionary<Parameter[]>): string[] {
...
-      def = `params.${group[0].name};`;
+      // when the schema: { '$ref': '#/definitions/exampleDto' } construct is used
+      if ('schema' in group[0]) {
+        def = `params.${group[0].name};`;
+      } else {
+        const list = _.map(group, p => `${p.name}: params.${p.name},`);
+        def = '{\n' + indent(list) + '\n};';
+      }
+
+      // when the patch method is used, it will discard bodyParams keys with value === undefined
```
Comment (jnwltr) [inline src/requests/process-method.ts]: Happens for all `body` params, not only for `patch`, does not it?

File: src/requests/process-method.ts (hunk 4 above)
Comment (fazpu) [reply src/requests/process-method.ts]: removed from getParamsSignature and getInterfaceDef

File: src/requests/process-method.ts (hunk 5 above)
Comment (fazpu) [reply src/requests/process-method.ts]: comment changed

**Pull request merged**